File: Python/polyamthread.py
import math, time
from multiprocessing.dummy import Pool as ThreadPool 

o=0
e=0
fac={}
def primes(n):
	global e
	global o
	ogn=n
	# print(n)
	if n%100==0:
		print(n)
	# if n == 1:	
	# 	e+=1
	primfac = []
	d = 2
	while d*d <= n:
		while (n % d) == 0:
			primfac.append(d)  # supposing you want multiple factors repeated
			n //= d
		d += 1
	if n > 1:
		primfac.append(n)

		# print(primfac)
	if len(primfac)%2==0:
		fac[ogn]='e'
		# e+=1
	else:
		fac[ogn]='o'
		# o+=1

pool = ThreadPool(1000) 
# results = pool.map(primes, range(1,1000))
results = pool.map(primes, range(1,100000000))
# results = pool.map(primes, range(100000001,200000000))
# results = pool.map(primes, range(200000001,300000000))
# results = pool.map(primes, range(300000001,400000000))
# results = pool.map(primes, range(400000001,500000000))
# results = pool.map(primes, range(500000001,600000000))
# results = pool.map(primes, range(600000001,700000000))
# results = pool.map(primes, range(700000001,800000000))
# results = pool.map(primes, range(800000001,900000000))
# results = pool.map(primes, range(900000001,1000000000))
for key, value in fac.items():
	# print(str(key)+" "+str(value))
	if value=='o':
		o+=1
	else:
		e+=1
	if(e>o):
		print("Polyas: False @ "+ str(key))
if (o>=e):
	print("Polyas: True")
print("Odd: "+str(o)+" Even: "+str(e))

File: C++/polyamthread.cpp
#include<iostream>
using std::cout; 
using std::endl;
#include <math.h>
#include<vector>
using std::vector;
#include<map>
using std::map;
#include<thread>
using std::thread;

static const int num_threads = 16;

map<long,char> m;

long o=0;
long e=0;
void primes(long n){
	long ogn;
	ogn = n;
	if (n%1000000==0 || n==906150256 || n==906150257){
		cout<<n<<endl;
	}
	vector<long> primfac;
	while (n%2 == 0)
    {
    	n = n/2;
        primfac.push_back(n);
    }
    for (int i = 3; i <= sqrt(n); i = i+2)
    {
        // While i divides n, print i and divide n
        while (n%i == 0)
        {
            primfac.push_back(i);
            n = n/i;
        }
    }
    if (n > 2){
	    primfac.push_back(n);
    }
	if (primfac.size()%2==0){
		m[ogn]='e';
		// e++;
	}
	else{
		m[ogn]='o';
		// o++;
	}
	// if(e>o && n!=1){
	// 	cout<<ogn<<" Winner Winner Chicken Dinner";
	// 	exit(0);
	// }
	}

int main(){
	cout<<thread::hardware_concurrency();
	thread t[num_threads];
	for (int i = 1; i < num_threads; ++i)
	{
		// for (int i = 0; i < count; ++i)
		// {
		// }
		// t[i] = thread(primes, j);	
	}
	for (int i = 1; i < num_threads; ++i) {
        t[i].join();
    }

	for(auto x:m){
		if (x.second=='e'){
			e++;
		}else{
			o++;
		}
	}
	if (o>=e){
		cout<<("Polyas: True")<<endl;
		cout<<"Odd: "<< o << endl << "Even: " << e<<endl;
	}
	return 0;
}

